### PR TITLE
Nonaffine swizzle formulation ep.2: Loop swizzle variant.

### PR DESCRIPTION
**Pull request by shmsong**
This PR introduces the concept of loop swizzle, and initial support to unblock matmul use cases more details on inlined comment. 

Remaining items:

- [x] Rebase and incorporate loop swizzle in matmul scheduler.
- [x] Add validation checks for loop swizzle.
- [ ] (After merging #1823, add block swizzle to the matmul parameters).

(In a follow up) pipe through loop swizzle in replay logic.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp [hunk 1]
```diff
@@ -1037,17 +1037,76 @@ void validateMma(Fusion* fusion) {
...
+
+      // Check not inlined swizzles:
+      //  Apply the loop swizzle check when it applies, and
+      // also make sure that the no swizzle is also in inlined_swizzle set.
+      // The latter would mean that one output of the swizzle is inlined while
+      //  the other is not. Such case will not be supported.
+      for (auto swizzle_expr : not_inlined_swizzles) {
+        TORCH_INTERNAL_ASSERT(
+            !inlined_swizzle_set.count(swizzle_expr),
+            "Cannot inline across swizzle expr ",
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_validation.cpp]: Nit: "Cannot partially inline across swizzle domains."

File: torch/csrc/jit/codegen/cuda/ir_internal_nodes.h [hunk 2]
```diff
@@ -1377,7 +1387,50 @@ class TORCH_CUDA_CU_API Swizzle2D : public Expr {
...
+  Swizzle2DType swizzle_type_ = Swizzle2DType::NoSwizzle;
+
+  // Swizzle mode of this swizzle instance.
+  // [Note on swizzle mode]
+  // On the current implementations we support two modes of
+  //  swizzle math, namely, data mode and loop mode.
+  // `Data` mode swizzling is a swizzle that will change the
+  //  data layout in shared memory, likely in global memory buffers
+  //  as well in the future and consumers of data-mode-swizzled buffers
+  //  will have to swizzle the read index accordingly. Most important
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_internal_nodes.h]: Just having trouble parsing `and consumers of data-mode-swizzled buffers will have to swizzle the read index accordingly.`

File: torch/csrc/jit/codegen/cuda/test/test_gpu.cpp [hunk 3]
```diff
@@ -23883,6 +23883,137 @@ TEST_F(NVFuserTest, FusionSwizzleMapping_CUDA) {
...
+  //[O, 4, 4]
+
+  // Swizzle the inner tile.
+  tv2->swizzle(Swizzle2DType::ZShape, -2, -1, SwizzleMode::Loop);
+
+  // Make swizzle output not a leaf domain.
+  tv2->merge(-2);
+
+  tv0->computeAt(tv2, -1);
+  fusion.print();
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/test/test_gpu.cpp]: Cleanup.

**Review by csarofeen — APPROVED** [review]
Nice cleanup in the tests.
LGTM

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp (hunk 1 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_validation.cpp]: Updated. Thanks.

File: torch/csrc/jit/codegen/cuda/test/test_gpu.cpp (hunk 3 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/test/test_gpu.cpp]: Updated. Thanks.

File: torch/csrc/jit/codegen/cuda/ir_internal_nodes.h (hunk 2 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/ir_internal_nodes.h]: Removed this sentence and redirected on the comment to `IndexSwizzle`.

This sentence is just saying the data stored in shared mem has a swizzled layout which makes the indexing different for the next stage consumers. The difference between the index math on `Tshared` in the `data swizzle` vs `loop swizzle` examples on the comment is an example.

**Pull request merged**